### PR TITLE
Implement EIP-7623 dual calldata pricing mechanism

### PR DESCRIPTION
**Pull request by Karrq**
- Add gas constants STANDARD_TOKEN_COST (4) and TOTAL_COST_FLOOR_PER_TOKEN (10)
- Add has_eip_7623 configuration flag to runtime Config struct
- Update record_transaction() to implement dual pricing formula when enabled
- Add calculate_calldata_tokens() helper function in utils
- Add comprehensive unit tests for EIP-7623 gas calculations

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: runtime/src/lib.rs [hunk 1]
```diff
@@ -294,6 +294,8 @@ pub struct Config {
 	pub estimate: bool,
 	/// Has EIP-6780. See [EIP-6780](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-6780.md)
 	pub has_eip_6780: bool,
+	/// Enable EIP-7623 calldata cost calculation. See [EIP-7623](https://eips.ethereum.org/EIPS/eip-7623)
+	pub has_eip_7623: bool,
```
Comment (manuelmauro) [inline runtime/src/lib.rs:298]: This is a Pectra feature. A new config `pectra()` should be added with the flag set to `true`, while all previous configs should have it set to `false`.

File: gasometer/src/utils.rs [hunk 2]
```diff
@@ -1,5 +1,14 @@
...
 
+/// Calculate the number of tokens in calldata according to EIP-7623
+/// tokens = zero_bytes + (nonzero_bytes * 4)
+#[allow(dead_code)]
+pub fn calculate_calldata_tokens(data: &[u8]) -> u64 {
+	let zero_bytes = data.iter().filter(|&&byte| byte == 0).count() as u64;
+	let nonzero_bytes = data.len() as u64 - zero_bytes;
+	zero_bytes + (nonzero_bytes * 4)
+}
+
```
Comment (manuelmauro) [inline gasometer/src/utils.rs]: Dead code. Should be removed.

File: gasometer/src/lib.rs [hunk 3]
```diff
@@ -1147,3 +1211,8 @@ impl MemoryCost {
 		}
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	mod eip_7623;
+}
```
Comment (manuelmauro) [inline gasometer/src/lib.rs]: Doesn't seem to add any value. Could be removed.

File: gasometer/src/tests/eip_7623.rs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+		// Test with empty data
+		assert_eq!(calculate_calldata_tokens(&[]), 0);
+	}
+
+	#[test]
+	fn test_eip_7623_call_cost() {
+		// Create configs with and without EIP-7623
+		let mut config_enabled = Config::london();
+		config_enabled.has_eip_7623 = true;
```
Comment (manuelmauro) [inline gasometer/src/tests/eip_7623.rs]: Should use `Config::pectra();`

File: gasometer/src/tests/eip_7623.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
+#[cfg(test)]
```
Comment (manuelmauro) [inline gasometer/src/tests/eip_7623.rs]: The tests look like snapshot tests at best, but the implementation looks hard to maintain.

**Comment by Karrq** [comment]
Claude says: Addressed in commit 64ffcb6. Added `Config::pectra()` method at runtime/src/lib.rs:529 with `has_eip_7623: true` set. This follows the established pattern for hard fork configurations and properly groups EIP-7623 with other Pectra features like EIP-6780, tloadstore, and mcopy.

@manuelmauro This addresses your comment about adding a pectra() config.

**Comment by Karrq** [comment]
Claude says: Addressed in commit c28006d. Removed the `calculate_calldata_tokens` function from gasometer/src/utils.rs. The token calculation logic is now inlined directly in the gasometer implementation where it's used, avoiding the dead code issue.

@manuelmauro This addresses your comment about removing the dead code.

**Comment by Karrq** [comment]
Claude says: Addressed in commit 6d49036. Simplified the test module structure by removing the unnecessary nested declaration. Tests remain in `src/tests/` directory with a proper `mod.rs` file, maintaining organization while eliminating the redundant wrapper.

@manuelmauro This addresses your comment about the test module structure not adding value.

**Comment by Karrq** [comment]
Claude says: Addressed in commit d643cb5. The tests have been significantly improved beyond simple snapshot tests:

**What the tests verify:**

1. **Floor vs Standard Cost Mechanism** (`test_eip_7623_call_cost_floor_mechanism`): Verifies that EIP-7623 correctly applies the max(standard_cost, floor_cost) formula. Tests with 1000 non-zero bytes showing floor cost (40,000 gas) wins over standard cost (16,000 gas).

2. **Token Calculation Properties** (`test_token_calculation_properties`): Validates the mathematical properties - zero bytes count as 1 token each, non-zero bytes count as 4 tokens each.

3. **Edge Cases** (`test_eip_7623_edge_cases`): Tests empty calldata (only base cost) and all-zero calldata patterns to ensure correct handling of boundary conditions.

4. **Mixed Data Patterns** (`test_eip_7623_mixed_data_patterns`): Verifies correct token counting for realistic data with mixed zero/non-zero bytes.

5. **Access List Integration** (`test_eip_7623_with_access_list`): Ensures access list costs are properly included in the standard cost calculation before comparing with floor cost.

6. **Create Transaction Costs** (`test_eip_7623_create_cost`): Validates that contract creation includes initcode costs in the standard cost calculation.

7. **Cost Comparison** (`test_eip_7623_cost_comparison`): Demonstrates when EIP-7623 increases costs compared to legacy pricing, particularly for small transactions.

**Potential test additions to consider:**

1. **Maximum calldata size test** - Test with calldata at block gas limit boundaries
2. **Gas exhaustion scenarios** - Verify behavior when calculated cost exceeds available gas  
3. **Interaction with other EIPs** - Test EIP-7623 with EIP-2930 access lists of varying sizes
4. **Performance regression test** - Ensure token calculation doesn't significantly impact gas calculation performance
5. **Fuzz testing** - Random calldata patterns to catch edge cases

The current test suite provides comprehensive coverage of the EIP-7623 mechanism with clear assertions and helper functions that make the tests maintainable and self-documenting.

@manuelmauro This addresses your comment about test quality and maintainability.

**Review by Karrq — COMMENTED** [review]
Claude says: I've addressed all the review comments in the following commits:

**Re: Config::pectra() comment** - Addressed in commit 64ffcb6. Added `Config::pectra()` method at runtime/src/lib.rs:529 with `has_eip_7623: true` set. This follows the established pattern for hard fork configurations and properly groups EIP-7623 with other Pectra features.

**Re: Dead code comment** - Addressed in commit c28006d. Removed the `calculate_calldata_tokens` function from gasometer/src/utils.rs. The token calculation logic is now inlined directly in the gasometer implementation.

**Re: Test module structure** - Addressed in commit 6d49036. Simplified the test module structure by removing the unnecessary nested declaration while maintaining organization in `src/tests/`.

**Re: Test quality** - Addressed in commit d643cb5. The tests now provide comprehensive coverage:

1. **Floor vs Standard Cost Mechanism** - Verifies max(standard_cost, floor_cost) formula with 1000 non-zero bytes showing floor cost wins
2. **Token Calculation Properties** - Validates zero bytes = 1 token, non-zero bytes = 4 tokens
3. **Edge Cases** - Tests empty and all-zero calldata patterns
4. **Mixed Data Patterns** - Verifies realistic data with mixed zero/non-zero bytes
5. **Access List Integration** - Ensures access list costs are included before floor comparison
6. **Create Transaction Costs** - Validates initcode costs in standard calculation
7. **Cost Comparison** - Demonstrates when EIP-7623 increases costs vs legacy

Potential future test additions: max calldata size, gas exhaustion scenarios, deeper EIP-2930 integration, performance regression tests, and fuzz testing.

File: runtime/src/lib.rs (hunk 1 above)
Comment (Karrq) [reply runtime/src/lib.rs:298]: Claude says: Addressed in commit 64ffcb6. Added `Config::pectra()` method at runtime/src/lib.rs:529 with `has_eip_7623: true` set. This follows the established pattern for hard fork configurations and properly groups EIP-7623 with other Pectra features like EIP-6780, tloadstore, and mcopy.

File: gasometer/src/utils.rs (hunk 2 above)
Comment (Karrq) [reply gasometer/src/utils.rs]: Claude says: Addressed in commit c28006d. Removed the `calculate_calldata_tokens` function from gasometer/src/utils.rs. The token calculation logic is now inlined directly in the gasometer implementation where it's used, avoiding the dead code issue.

File: gasometer/src/lib.rs (hunk 3 above)
Comment (Karrq) [reply gasometer/src/lib.rs]: Claude says: Addressed in commit 6d49036. Simplified the test module structure by removing the unnecessary nested declaration. Tests remain in `src/tests/` directory with a proper `mod.rs` file, maintaining organization while eliminating the redundant wrapper.

File: gasometer/src/tests/eip_7623.rs (hunk 5 above)
Comment (Karrq) [reply gasometer/src/tests/eip_7623.rs]: Claude says: Addressed in commit d643cb5. The tests have been significantly improved beyond simple snapshot tests:

**What the tests verify:**

1. **Floor vs Standard Cost Mechanism** (`test_eip_7623_call_cost_floor_mechanism`): Verifies that EIP-7623 correctly applies the max(standard_cost, floor_cost) formula. Tests with 1000 non-zero bytes showing floor cost (40,000 gas) wins over standard cost (16,000 gas).

2. **Token Calculation Properties** (`test_token_calculation_properties`): Validates the mathematical properties - zero bytes count as 1 token each, non-zero bytes count as 4 tokens each.

3. **Edge Cases** (`test_eip_7623_edge_cases`): Tests empty calldata (only base cost) and all-zero calldata patterns to ensure correct handling of boundary conditions.

4. **Mixed Data Patterns** (`test_eip_7623_mixed_data_patterns`): Verifies correct token counting for realistic data with mixed zero/non-zero bytes.

5. **Access List Integration** (`test_eip_7623_with_access_list`): Ensures access list costs are properly included in the standard cost calculation before comparing with floor cost.

6. **Create Transaction Costs** (`test_eip_7623_create_cost`): Validates that contract creation includes initcode costs in the standard cost calculation.

7. **Cost Comparison** (`test_eip_7623_cost_comparison`): Demonstrates when EIP-7623 increases costs compared to legacy pricing, particularly for small transactions.

**Potential test additions to consider:**
- Maximum calldata size test - Test with calldata at block gas limit boundaries
- Gas exhaustion scenarios - Verify behavior when calculated cost exceeds available gas
- Interaction with other EIPs - Test EIP-7623 with EIP-2930 access lists of varying sizes
- Performance regression test - Ensure token calculation doesn't significantly impact gas calculation performance
- Fuzz testing - Random calldata patterns to catch edge cases

The current test suite provides comprehensive coverage of the EIP-7623 mechanism with clear assertions and helper functions that make the tests maintainable and self-documenting.

File: gasometer/src/lib.rs [hunk 6]
```diff
@@ -248,25 +250,53 @@ impl<'config> Gasometer<'config> {
...
-					access_list_storage_len
-				);
-
-				cost
+				if self.config.has_eip_7623 {
+					// EIP-7623: Dual pricing mechanism
+					let tokens_in_calldata = zero_data_len as u64 + (non_zero_data_len as u64 * 4);
+					let access_list_cost = access_list_address_len as u64
+						* self.config.gas_access_list_address
+						+ access_list_storage_len as u64 * self.config.gas_access_list_storage_key;
```
Comment (RomarQ) [inline gasometer/src/lib.rs]: We should use safe math here, like `saturating_add`, `saturating_mul`

File: gasometer/src/lib.rs (hunk 3 above)
Comment (manuelmauro) [reply gasometer/src/lib.rs]: This is still weird. Rust has its own test folders, alternatively use in-module tests.

File: gasometer/src/tests/eip_7623.rs (hunk 4 above)
Comment (manuelmauro) [reply gasometer/src/tests/eip_7623.rs]: Addressed.

File: gasometer/src/lib.rs (hunk 6 above)
Comment (Karrq) [reply gasometer/src/lib.rs]: Claude says: Addressed in commit 281b86e. I've replaced all unchecked arithmetic operations with their checked equivalents (checked_mul, checked_add) that return ExitError::OutOfGas on overflow. This follows the established pattern in the codebase for gas calculations and ensures safety in this consensus-critical code.

File: gasometer/src/lib.rs (hunk 3 above)
Comment (Karrq) [reply gasometer/src/lib.rs]: Claude says: Addressed in commit 900b17f. I've moved the tests from the separate src/tests/ directory to an inline #[cfg(test)] module within lib.rs. This follows the more idiomatic Rust pattern for unit tests as you suggested.
